gufe/protocols/protocolunit.py

```diff
@@ -7,79 +7,312 @@
+
+    def __repr__(self):
+        return f"{type(self).__name__}({self.name})"
+
+    def _gufe_tokenize(self):
+        if self._pure:
+            return normalize(self.to_dict(include_defaults=False))
+        else:
+            # tokenize with uuid
+            return uuid.uuid4()
+
+    def _defaults(self):
+        # not used by `ProtocolDAG`
+        return {}
+
+    def _to_dict(self):
+        return {'name': self.name,
+                'source_key': self.source_key,
+                'pure': self.pure,
+                'inputs': self.inputs,
+                'outputs': self.outputs}
+
+    @classmethod
+    def _from_dict(cls, dct: Dict):
+        return cls(**dct)
+
+    @property
+    def name(self):
+        """
+
+        """
+        # set name to source_key if not manually set; used for display
+        if self._name is None:
+            self._name = self.source_key
+        return self._name
+
+    @property
+    def source_key(self):
+        return self._source_key
+
+    @property
+    def pure(self):
+        return self._pure
+
+    @property
+    def inputs(self):
+        return self._inputs
+
+    @property
+    def outputs(self):
+        return self._outputs
+
+
+class ProtocolUnitResult(ProtocolUnitResultBase):
+    """Result for a single `ProtocolUnit` execution.
+
+    Immutable upon creation.
+
+    """
+
+    def ok(self) -> bool:
+        return True
+
+
+class ProtocolUnitFailure(ProtocolUnitResultBase):
+    exception: Exception
+
+    def __init__(self, *, name=None, source_key, pure, inputs, outputs, exception, traceback):
+        self._exception = exception
+        self._traceback = traceback
+        super().__init__(name=name, source_key=source_key, pure=pure, inputs=inputs, outputs=outputs)
+
+    @property
+    def exception(self):
+        return self._exception
+
+    @property
+    def traceback(self):
+        return self._traceback
+
+    def ok(self) -> bool:
+        return False
+
+
+class ProtocolUnit(GufeTokenizable):
+    """A unit of work within a ProtocolDAG."""
 
     def __init__(
         self,
-        settings: Optional["ProtocolSettings"] = None,  # type: ignore
+        *,
         name: Optional[str] = None,
-        **kwargs
+        pure: bool = False,
+        **inputs
     ):
-        self._settings = settings
-        self._kwargs = kwargs
+        """Create an instance of a ProtocolUnit.
+
+        Parameters
+        ----------
+        name : str
+            Custom name to give this 
+        **inputs 
+            Keyword arguments, which an include other `ProtocolUnit`s on which this
+            `ProtocolUnit` is dependent. For serializability, should be composed of
+
+        """
         self._name = name
+        self._pure = pure
+        self._inputs = inputs
 
-    def __hash__(self):
-        return hash(
-            (self.__class__.__name__, self._settings, frozenset(self._kwargs.items()))
-        )
+        self.dag_scratch = None
```

Comment:
Why set these as attributes? Just have them as variables in execute and they are persisted in the Results

Reply:
Ah, apologies, missed removing these from an earlier iteration. These now make their way into `ProtocolUnit._execute` via `ctx`, a `Context` object.